_posts/2020-02-03-apache-beam-how-beam-runs-on-top-of-flink.md

```diff
@@ -0,0 +1,157 @@
+
+### AssignWindows
+
+In Flink, windows are assigned by the `WindowOperator` when you use the `window()` in the API. In Beam, windows can be assigned at any point in time. Any element is implicitly part of a window. If no window is assigned explicitly, it is part of the `GlobalWindow`. Window information is stored for each element in `WindowedValue`. The window information is only useful once we issue a `GroupByKey`.
+
+### GroupByKey
+
+Most of the time it is useful to partition the data by a key. In Flink, this is done via the `keyBy()` API call. In Beam the `GroupByKey` transform can only be applied if the input is of form `KV<Key, Value>`. Unlike Flink where the key can even be nested inside the data, Beam enforces the key to always be explicit. The GroupByKey transform then groups the data by key and by window which is similar to what `keyBy(..).window(..)` would give us in Flink. Beam has its own set of libraries to do that because Beam has its own set of window function and triggers. Essentially, it is very similar to what the WindowOperator does in Flink.
+
+### Flatten
+
+The Flatten operator takes multiple DataSet/DataStreams, called P[arallel]Collections in Beam, and combines them into one collection. This is equivalent to Flink's `union()` operation.
+
+
+## The “Portable” Flink Runner in Beam
+
+The portable Flink Runner in Beam is the evolution of the classic Runner. Classic Runners are tied to the JVM ecosystem, but the Beam community wanted to move past this and also execute Python, Go and other languages. This adds another dimension to Beam in terms of portability because, like previously mentioned, Beam already had portability across execution engines. It was necessary to change the translation logic of the Runner to be able to support language portability.
+
+There are two important building blocks for portable Runners: 
+1. A common pipeline format across all the languages: The Runner API
+2. A common interface during execution for the communication between the Runner and the code written in any language: The Fn API
+
+The Runner API provides a universal representation of the pipeline as Protobuf which contains the transforms, types, and the user code. Protobuf was chosen as the format because every language has libraries available for it. Similarly, for the execution part, Beam introduced the Fn API interface to handle the communication between the Runner/execution engine and the user code that may be written in a different language and executes in a different process. Fn API is pronounced "fun API", you may guess why.
```

Comment:
```suggestion
The Runner API provides a universal representation of the pipeline as Protobuf which contains the transforms, types, and user code. Protobuf was chosen as the format because every language has libraries available for it. Similarly, for the execution part, Beam introduced the Fn API interface to handle the communication between the Runner/execution engine and the user code that may be written in a different language and executes in a different process. Fn API is pronounced "fun API", you may guess why.
```